In [1]:
import pandas as pd
import numpy as np

In [2]:
# reading the data
ev_df = pd.read_csv("../data/Electric_Vehicle_Population_Data.csv")
ev_df.head()

,VIN (1-10),County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Base MSRP,Legislative District,DOL Vehicle ID,Vehicle Location,Electric Utility,2020 Census Tract
0,5YJYGDEE1L,King,Seattle,WA,98122.0,2020,TESLA,MODEL Y,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,291,0,37.0,125701579,POINT (-122.30839 47.610365),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),5.303301e+10
1,7SAYGDEE9P,Snohomish,Bothell,WA,98021.0,2023,TESLA,MODEL Y,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,0,0,1.0,244285107,POINT (-122.179458 47.802589),PUGET SOUND ENERGY INC,5.306105e+10
2,5YJSA1E4XK,King,Seattle,WA,98109.0,2019,TESLA,MODEL S,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,270,0,36.0,156773144,POINT (-122.34848 47.632405),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),5.303301e+10
3,5YJSA1E27G,King,Issaquah,WA,98027.0,2016,TESLA,MODEL S,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,210,0,5.0,165103011,POINT (-122.03646 47.534065),PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA),5.303303e+10
4,5YJYGDEE5M,Kitsap,Suquamish,WA,98392.0,2021,TESLA,MODEL Y,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,0,0,23.0,205138552,POINT (-122.55717 47.733415),PUGET SOUND ENERGY INC,5.303594e+10


## Cleaning the column names

In [3]:
import re

ev_df.columns=(
    ev_df.columns
    .str.strip()
    .str.lower()
    .str.replace(r'[^\w\s]', '', regex=True) # remove punctuation
    .str.replace(r'\s+', '_', regex=True) # replace spaces with underscores
)
ev_df.head()

,vin_110,county,city,state,postal_code,model_year,make,model,electric_vehicle_type,clean_alternative_fuel_vehicle_cafv_eligibility,electric_range,base_msrp,legislative_district,dol_vehicle_id,vehicle_location,electric_utility,2020_census_tract
0,5YJYGDEE1L,King,Seattle,WA,98122.0,2020,TESLA,MODEL Y,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,291,0,37.0,125701579,POINT (-122.30839 47.610365),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),5.303301e+10
1,7SAYGDEE9P,Snohomish,Bothell,WA,98021.0,2023,TESLA,MODEL Y,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,0,0,1.0,244285107,POINT (-122.179458 47.802589),PUGET SOUND ENERGY INC,5.306105e+10
2,5YJSA1E4XK,King,Seattle,WA,98109.0,2019,TESLA,MODEL S,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,270,0,36.0,156773144,POINT (-122.34848 47.632405),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),5.303301e+10
3,5YJSA1E27G,King,Issaquah,WA,98027.0,2016,TESLA,MODEL S,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,210,0,5.0,165103011,POINT (-122.03646 47.534065),PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA),5.303303e+10
4,5YJYGDEE5M,Kitsap,Suquamish,WA,98392.0,2021,TESLA,MODEL Y,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,0,0,23.0,205138552,POINT (-122.55717 47.733415),PUGET SOUND ENERGY INC,5.303594e+10


In [4]:
# checking info
ev_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 177866 entries, 0 to 177865
Data columns (total 17 columns):
 #   Column                                           Non-Null Count   Dtype  
---  ------                                           --------------   -----  
 0   vin_110                                          177866 non-null  object 
 1   county                                           177861 non-null  object 
 2   city                                             177861 non-null  object 
 3   state                                            177866 non-null  object 
 4   postal_code                                      177861 non-null  float64
 5   model_year                                       177866 non-null  int64  
 6   make                                             177866 non-null  object 
 7   model                                            177866 non-null  object 
 8   electric_vehicle_type                            177866 non-null  object 
 9   clean_alternati

## Checking for missing values

In [5]:
missing_ev = ev_df.isnull().sum().sort_values(ascending=True)
print("Missing Values:\n", missing_ev)

Missing Values:
 vin_110                                              0
dol_vehicle_id                                       0
base_msrp                                            0
electric_range                                       0
clean_alternative_fuel_vehicle_cafv_eligibility      0
model                                                0
electric_vehicle_type                                0
model_year                                           0
state                                                0
make                                                 0
electric_utility                                     5
postal_code                                          5
city                                                 5
county                                               5
2020_census_tract                                    5
vehicle_location                                     9
legislative_district                               389
dtype: int64


### Checking unique values in `legislative_district`

In [10]:
unique_ld = ev_df['legislative_district'].dropna().unique()
print(sorted(unique_ld))

[np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0), np.float64(5.0), np.float64(6.0), np.float64(7.0), np.float64(8.0), np.float64(9.0), np.float64(10.0), np.float64(11.0), np.float64(12.0), np.float64(13.0), np.float64(14.0), np.float64(15.0), np.float64(16.0), np.float64(17.0), np.float64(18.0), np.float64(19.0), np.float64(20.0), np.float64(21.0), np.float64(22.0), np.float64(23.0), np.float64(24.0), np.float64(25.0), np.float64(26.0), np.float64(27.0), np.float64(28.0), np.float64(29.0), np.float64(30.0), np.float64(31.0), np.float64(32.0), np.float64(33.0), np.float64(34.0), np.float64(35.0), np.float64(36.0), np.float64(37.0), np.float64(38.0), np.float64(39.0), np.float64(40.0), np.float64(41.0), np.float64(42.0), np.float64(43.0), np.float64(44.0), np.float64(45.0), np.float64(46.0), np.float64(47.0), np.float64(48.0), np.float64(49.0)]


### Checking how many times each `legislative_district` appears

In [11]:
print(ev_df['legislative_district'].value_counts(dropna=False).sort_index())

legislative_district
1.0      7786
2.0      2015
3.0       876
4.0      1470
5.0      7614
6.0      1691
7.0       839
8.0      1831
9.0      1003
10.0     3018
11.0     7507
12.0     1644
13.0     1211
14.0     1123
15.0      443
16.0      885
17.0     3000
18.0     4744
19.0     1032
20.0     1582
21.0     4410
22.0     4399
23.0     3830
24.0     2391
25.0     1838
26.0     3538
27.0     2522
28.0     2277
29.0     1183
30.0     2120
31.0     3100
32.0     4358
33.0     3550
34.0     5155
35.0     2488
36.0     7389
37.0     5215
38.0     1736
39.0     2593
40.0     3798
41.0    11553
42.0     2444
43.0     6524
44.0     4677
45.0    10737
46.0     6901
47.0     3230
48.0     9817
49.0     2390
NaN       389
Name: count, dtype: int64


The above result provides no conclusion in order to impute `legislative_district`

### Checking how consistent `postal_code` is with `legislative_district`

In [12]:
postal_district_counts  = ev_df.dropna(subset=['legislative_district']).groupby('postal_code')['legislative_district'].nunique()
print(postal_district_counts.value_counts())

legislative_district
1    356
2    145
3     36
4      6
Name: count, dtype: int64


### Imputing the NaNs using `postal_code` with a 1-to-1 district mapping

In [14]:
# creating postal to district mapping
postal_to_district = ev_df.dropna(subset=["legislative_district"]).groupby("postal_code")["legislative_district"].agg(lambda x: x.mode().iloc[0] if x.nunique() == 1 else pd.NA)

# filling missing values using the above mapping
ev_df["legislative_district"] = ev_df.apply(
    lambda row: postal_to_district[row["postal_code"]] if pd.isna(row["legislative_district"]) and row["postal_code"] in postal_to_district and pd.notna(postal_to_district[row["postal_code"]]) else row["legislative_district"],
    axis = 1
)

In [16]:
ev_df['legislative_district'].isna().sum()

np.int64(389)

In [17]:
missing_district = ev_df[ev_df['legislative_district'].isna()]
print(missing_district['postal_code'].isin(postal_to_district.index).sum())

0


### Debugging

In [22]:
missing_district = ev_df[ev_df['legislative_district'].isna()]

# how many have NaN postal codes
print(missing_district['postal_code'].isna().sum())

# how many postal codes are mappable
print(missing_district['postal_code'].isin(postal_to_district.index).sum())  

0
0


The result indicates that all 389 postal codes are ambiguous (many-to-one or many-to-many mappings)

In [24]:
# Create mapping only for cities with 1-to-1 district mapping
city_to_district = ev_df.dropna(subset=['legislative_district']) \
    .groupby('city')['legislative_district'] \
    .agg(lambda x: x.mode().iloc[0] if x.nunique() == 1 else pd.NA)

In [25]:
# Ensure data types match
ev_df['city'] = ev_df['city'].astype(str)
city_to_district.index = city_to_district.index.astype(str)

In [26]:
# Impute remaining NaNs using city mapping
ev_df['legislative_district'] = ev_df.apply(
    lambda row: city_to_district[row['city']] if pd.isna(row['legislative_district']) and row['city'] in city_to_district and pd.notna(city_to_district[row['city']]) else row['legislative_district'],
    axis=1
)

In [27]:
ev_df['legislative_district'].isna().sum()

np.int64(383)

The above result indicates that city-based fallback successfully imputed 6 more values!

**What this tells is:**
- The majority of the remaining `NaNs` are:
  - From postal codes with ambiguous mappings, and
  - From cities that also map to multiple districts.

### Labeling the remaining as `Missing`

In [28]:
ev_df['legislative_district'] = ev_df['legislative_district'].fillna('Missing').astype(str)
ev_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 177866 entries, 0 to 177865
Data columns (total 17 columns):
 #   Column                                           Non-Null Count   Dtype  
---  ------                                           --------------   -----  
 0   vin_110                                          177866 non-null  object 
 1   county                                           177861 non-null  object 
 2   city                                             177866 non-null  object 
 3   state                                            177866 non-null  object 
 4   postal_code                                      177866 non-null  object 
 5   model_year                                       177866 non-null  int64  
 6   make                                             177866 non-null  object 
 7   model                                            177866 non-null  object 
 8   electric_vehicle_type                            177866 non-null  object 
 9   clean_alternati

In [29]:
# View rows with any missing values except legislative_district
missing_cols = ['county', 'electric_utility', '2020_census_tract', 'vehicle_location']
ev_df[ev_df[missing_cols].isnull().any(axis=1)]

,vin_110,county,city,state,postal_code,model_year,make,model,electric_vehicle_type,clean_alternative_fuel_vehicle_cafv_eligibility,electric_range,base_msrp,legislative_district,dol_vehicle_id,vehicle_location,electric_utility,2020_census_tract
108615,JTDKARFP1H,Pacific,Long Beach,WA,98634.0,2017,TOYOTA,PRIUS PRIME,Plug-in Hybrid Electric Vehicle (PHEV),Not eligible due to low battery range,25,0,19.0,171247282,NaN,BONNEVILLE POWER ADMINISTRATION||PUD NO 2 OF P...,5.304995e+10
149245,5YJSA1E21J,NaN,nan,AE,nan,2018,TESLA,MODEL S,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,249,0,Missing,220643132,NaN,NaN,NaN
154553,5YJ3E1EA5K,NaN,nan,BC,nan,2019,TESLA,MODEL 3,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,220,0,Missing,475254825,NaN,NaN,NaN
159157,5YJRE11B48,NaN,nan,BC,nan,2008,TESLA,ROADSTER,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,220,98950,Missing,143609049,NaN,NaN,NaN
159987,KMHE54L2XH,Pacific,Long Beach,WA,98634.0,2017,HYUNDAI,SONATA,Plug-in Hybrid Electric Vehicle (PHEV),Not eligible due to low battery range,27,0,19.0,245533578,NaN,BONNEVILLE POWER ADMINISTRATION||PUD NO 2 OF P...,5.304995e+10
165250,1G1FW6S08N,Pacific,Long Beach,WA,98634.0,2022,CHEVROLET,BOLT EV,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,0,0,19.0,218102209,NaN,BONNEVILLE POWER ADMINISTRATION||PUD NO 2 OF P...,5.304995e+10
167298,5YJXCAE24H,NaN,nan,BC,nan,2017,TESLA,MODEL X,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,200,0,Missing,159850029,NaN,NaN,NaN
170413,1N4AZ0CP1D,NaN,nan,AP,nan,2013,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,75,0,Missing,112277063,NaN,NaN,NaN
173101,1C4JJXP69P,Pacific,Long Beach,WA,98634.0,2023,JEEP,WRANGLER,Plug-in Hybrid Electric Vehicle (PHEV),Not eligible due to low battery range,21,0,19.0,254934886,NaN,BONNEVILLE POWER ADMINISTRATION||PUD NO 2 OF P...,5.304995e+10


The above result confirms that all 5 missing `county`, `electric_utility`, `2020_census_tract` values are part of the same 9 rows that also have missing `vehicle_location`.

### Dropping the 9 rows

In [30]:
# Drop rows where any of the 4 important location-related columns are missing
ev_df = ev_df.dropna(subset=['county', 'electric_utility', '2020_census_tract', 'vehicle_location'])
ev_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 177857 entries, 0 to 177865
Data columns (total 17 columns):
 #   Column                                           Non-Null Count   Dtype  
---  ------                                           --------------   -----  
 0   vin_110                                          177857 non-null  object 
 1   county                                           177857 non-null  object 
 2   city                                             177857 non-null  object 
 3   state                                            177857 non-null  object 
 4   postal_code                                      177857 non-null  object 
 5   model_year                                       177857 non-null  int64  
 6   make                                             177857 non-null  object 
 7   model                                            177857 non-null  object 
 8   electric_vehicle_type                            177857 non-null  object 
 9   clean_alternative_fu

In [6]:
# check for duplicates
duplicates_ev = ev_df.duplicated().sum()
print(f"Duplicate rows: {duplicates_ev}")

Duplicate rows: 0
